File: leetcode/1-100/77-combinations/bench/combinations.c
/*
 * Benchmark workload — Combinations (LeetCode #77).
 * C mirror of bench/combinations.kara. Start-indexed pruned backtracking that
 * ENUMERATES all C(16,8)=12870 combinations and folds each leaf's values into a
 * threaded accumulator (no storage), K=1500 iterations seeded by the iteration
 * index. The path is a small stack buffer indexed by depth (equivalent to Kara's
 * reused Vec push/pop). The DFS recursion is the measured work. See ../README.md.
 */
#include <stdio.h>
#include <stdint.h>

static int64_t enumerate(int64_t n, int64_t k, int64_t start,
                         int64_t *path, int64_t depth, int64_t acc) {
    if (depth == k) {
        int64_t a = acc;
        for (int64_t j = 0; j < k; j++)
            a = (a * 131 + path[j]) % 1000000007;
        return a;
    }
    int64_t need = k - depth;
    int64_t limit = n - need + 1;
    int64_t a = acc;
    for (int64_t i = start; i <= limit; i++) {
        path[depth] = i;
        a = enumerate(n, k, i + 1, path, depth + 1, a);
    }
    return a;
}

int main(void) {
    const int64_t n = 16, k = 8, total = 1500, modulus = 1000000007;
    int64_t path[64];
    int64_t sum = 0;
    for (int64_t iter = 0; iter < total; iter++) {
        int64_t r = enumerate(n, k, 1, path, 0, iter);
        sum = (sum + r) % modulus;
    }
    printf("%lld\n", (long long)sum);
    return 0;
}

File: leetcode/1-100/77-combinations/bench/combinations.rs
// Benchmark workload — Combinations (LeetCode #77).
// Rust mirror of bench/combinations.kara. Start-indexed pruned backtracking that
// ENUMERATES all C(16,8)=12870 combinations and folds each leaf's values into a
// threaded accumulator (no Vec-of-Vec storage), K=1500 iterations seeded by the
// iteration index. The DFS recursion is the measured work. See ../README.md.

fn enumerate(n: i64, k: i64, start: i64, path: &mut Vec<i64>, acc: i64) -> i64 {
    if path.len() as i64 == k {
        let mut a = acc;
        for &x in path.iter() {
            a = (a * 131 + x) % 1_000_000_007;
        }
        return a;
    }
    let need = k - path.len() as i64;
    let limit = n - need + 1;
    let mut a = acc;
    let mut i = start;
    while i <= limit {
        path.push(i);
        a = enumerate(n, k, i + 1, path, a);
        path.pop();
        i += 1;
    }
    a
}

fn main() {
    const N: i64 = 16;
    const K: i64 = 8;
    const TOTAL: i64 = 1500;
    const MODULUS: i64 = 1_000_000_007;

    let mut path: Vec<i64> = Vec::new();
    let mut sum: i64 = 0;
    for iter in 0..TOTAL {
        let r = enumerate(N, K, 1, &mut path, iter);
        sum = (sum + r) % MODULUS;
    }
    println!("{}", sum);
}
